docs/development.md

```diff
@@ -56,6 +56,139 @@ git push -f origin myfeature
 
 We follow Test Driven Development(TDD) workflow in our development process. You can read more about it [here](/docs/tdd-workflow.md).
 
+### Writing Unit tests
+
+#### Unit test for functions consuming client-go functions
+
+Unit-tests for the functions making API calls with client-go library 
+can be written by using package (fake)[https://godoc.org/k8s.io/client-go/kubernetes/fake].
+
+The way to mock the api calls is by mocking the actual api calls with functions defined in
+(client-go/testing)[https://godoc.org/k8s.io/client-go/testing]
+and using (pkg/testing)[https://golang.org/pkg/testing/] for writing the test.
+
+
+##### How to write unit tests having API calls in a nutshell?
+
+1. Identify the API calls being made by the function during the execution
+
+2. Initialise the relevant clientsets and clients 
+
+3. In case of API calls which are returning any object 
+  which is later being processed inside the function, 
+  Implement fake functions and use them instead.
+  Use addreactor method for adding corresponding clientset ref:(here)[https://godoc.org/k8s.io/client-go/testing#Fake.AddReactor].
+
+4. Use https://godoc.org/k8s.io/client-go/testing#Fake.Actions 
+  for validating number of fake actions performed and the values with which the fake calls were made.
+
+##### Example of using fake for testing functions making API calls.
+
+Initialising a fakeclientset and fakeclient properly is the first thing. 
+
+For example : 
+take the example of unit test for CreateRoute function in pkg/occlient/occlient.go 
+```
+/ CreateRoute creates a route object for the given service and with the given
+// labels
+func (c *Client) CreateRoute(name string, serviceName string, labels map[string]string) (*routev1.Route, error) {
+	route := &routev1.Route{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:   name,
+			Labels: labels,
+		},
+		Spec: routev1.RouteSpec{
+			To: routev1.RouteTargetReference{
+				Kind: "Service",
+				Name: serviceName,
+			},
+		},
+	}
+	r, err := c.routeClient.Routes(c.namespace).Create(route)
+	if err != nil {
+		return nil, errors.Wrap(err, "error creating route")
+	}
+	return r, nil
+}
+```
+
+Looking at the function body and identify how many API calls it is making while execution. 
```

Comment:
It can be observed that, the function above is making 1 api call as in the line below:

```
r, err := c.routeClient.Routes(c.namespace).Create(route)

```